File: engine/execution.py
"""
Zugausführung für Mosaic-AI.

Führt einen bereits validierten Move auf dem GameState aus.
Reihenfolge:
  1. Steine nehmen (Take) — Fabrik aktualisieren, Marker vergeben
  2. Steine legen (Place) — Musterreihe befüllen, Überschuss → Strafleiste
  3. Startspieler-Marker verarbeiten
  4. Bonus-Chip aufdecken falls Fabrik jetzt leer

WICHTIG: execute_move() geht davon aus dass der Move bereits validiert
wurde. Ungültige Züge führen zu undefiniertem Verhalten.
"""

from __future__ import annotations
from typing import TYPE_CHECKING

from engine.moves import Move, TakeSource
from engine.tile import TileColor

if TYPE_CHECKING:
    from engine.setup import GameState


def execute_move(state: "GameState", move: Move) -> None:
    """
    Führt einen validierten Zug aus und aktualisiert den GameState.
    """
    is_global_moon_take = (
        move.take.source.name == "SMALL_FACTORY_MOON" and 
        move.take.factory_id is None
    )
    
    if is_global_moon_take:
        _execute_moon_take(state, move.take.color, move.place.row_index)
        return  # Fertig! Der Rest von execute_move wird übersprungen.
    
    # 1. Steine nehmen
    take_result = _execute_take(state, move)
    if len(take_result) == 4:
        tiles, got_marker, moon_log, chip_log = take_result
    else:
        tiles, got_marker = take_result
        moon_log, chip_log = None, None

    # 2. Startspieler-Marker
    if got_marker:
        _apply_first_player_marker(state)

    # 3. Logging VOR _execute_place — damit Aktions-Log vor Strafleisten-Warnung erscheint
    player = state.active_player
    color = move.take.color
    dest = f"Reihe {move.place.row_index+1}" if move.place.row_index >= 0 else "Strafleiste"
    src_map = {
        "SMALL_FACTORY_SUN": f"F{move.take.factory_id}",
        "LARGE_FACTORY_SUN": "GF",
    }
    src_str = src_map.get(move.take.source.name, str(move.take.source))
    # 4. Aktions-Log + Place + Mondstapel/Chip-Logs
    ri = move.place.row_index
    # Aktions-Log VOR place — Index merken um Füllstand später einzufügen
    state.log_event(
        f"☀️  {player.name}: {len(tiles)}× {color.value} "
        f"von {src_str} → {dest}"
    )
    action_log_idx = len(state.log) - 1
    _execute_place(state, tiles, move.take.color, ri)

    # Füllstand direkt in den Aktions-Log-Eintrag einbauen (nur bei Musterreihe)
    if ri >= 0:
        row = player.pattern_lines[ri]
        state.log[action_log_idx] += f" [{len(row.tiles)}/{row.capacity}]"

    # 5. Mondstapel-Log und Chip-Aufdeckung
    if moon_log:
        state.log_event(moon_log)
    if chip_log:
        state.log_event(chip_log)

def _execute_moon_take(state, color, row_index: int) -> None:
    """
    Sonderaktion (Aktion C): Nimmt alle obersten Fliesen der gewählten Farbe
    vom Mondbereich ALLER Manufakturen (klein + groß) gleichzeitig.
    """
    p = state.active_player
    taken = []

    # 1. Kleine Fabriken abräumen — pro Fabrik tracken
    sources = []  # [(label, count), ...]
    pending_logs = []  # Mond-Stapel und Chip-Logs — erst nach Aktions-Log schreiben
    for f in state.factories:
        if color in f.moon_top_colors():
            tiles_from_f = f.take_from_moon(color)
            if tiles_from_f:
                taken += tiles_from_f
                sources.append((f"F{f.factory_id}", len(tiles_from_f)))
            # Mondstapel nach Entnahme sammeln — auch wenn leer
            stack_str = _format_moon_stacks(f) if f.moon_stacks else "leer"
            pending_logs.append(
                f"🌙 F{f.factory_id} Mond-Stapel: {stack_str}"
            )
            # Bonus-Chip aufdecken wenn Fabrik jetzt komplett leer
            if f.is_fully_empty and f.bonus_chip and not f.bonus_chip_revealed:
                f.bonus_chip_revealed = True
                pending_logs.append(f"🎴 F{f.factory_id}: Bonusplättchen aufgedeckt!")

    # 2. Große Manufaktur (Moon) abräumen
    got_marker = False
    if color in state.large_factory.moon_colors():
        tiles, got_marker = state.large_factory.take_from_moon(color)
        if tiles:
            taken += tiles
            sources.append(("GF", len(tiles)))
            # GF Pool nach Entnahme loggen
            pool = state.large_factory.moon_pool
            pool_str = ", ".join(c.value for c in pool) if pool else ""
            pool_display = f"({pool_str})" if pool else "leer"
            pending_logs.append(f"🌙 GF Moon-Pool: {pool_display}")

    # 3. Sicherheitscheck
    if not taken:
        raise ValueError(f"Keine {color.value}-Fliesen oben auf Moon-Seiten gefunden.")

    # 4. Startspieler-Marker vergeben
    if got_marker:
        _apply_first_player_marker(state)

    # 5. Aktions-Log + Place + Mondstapel/Chip-Logs
    dest = f"Reihe {row_index+1}" if row_index >= 0 else "Strafleiste"
    src_detail = "+".join(str(c) for _, c in sources)
    src_labels = ", ".join(lbl for lbl, _ in sources)

    # Aktions-Log VOR place — Index merken um Füllstand später einzufügen
    state.log_event(
        f"🌙 {p.name}: {len(taken)} ({src_detail})× {color.value} "
        f"von {src_labels} → {dest}"
    )
    action_log_idx = len(state.log) - 1
    _execute_place(state, taken, color, row_index)

    # Füllstand direkt in den Aktions-Log-Eintrag einbauen (nur bei Musterreihe)
    if row_index >= 0:
        row = p.pattern_lines[row_index]
        state.log[action_log_idx] += f" [{len(row.tiles)}/{row.capacity}]"

    for log in pending_logs:
        state.log_event(log)

# ---------------------------------------------------------------------------
# Take
# ---------------------------------------------------------------------------

def _execute_take(state: "GameState", move: Move) -> tuple[list[TileColor], bool]:
    """
    Führt den Take-Teil aus.
    Gibt (genommene_steine, hat_marker_bekommen) zurück.
    """
    src = move.take.source
    color = move.take.color

    if src == TakeSource.SMALL_FACTORY_SUN:
        return _take_small_sun(state, move)

    if src == TakeSource.SMALL_FACTORY_MOON:
        return _take_small_moon(state, move)

    if src == TakeSource.LARGE_FACTORY_SUN:
        return _take_large_sun(state, move)

    if src == TakeSource.LARGE_FACTORY_MOON:
        return _take_large_moon(state, move)

    raise ValueError(f"Unbekannte TakeSource: {src}")



def _format_moon_stacks(f) -> str:
    """Formatiert alle Mondstapel einer Fabrik als lesbaren String.
    Jeder Stapel: [unten → oben], sichtbar = letzter Stein.
    """
    if not f.moon_stacks:
        return "leer"
    stacks = []
    for stack in f.moon_stacks:
        top = stack[-1].value if stack else "?"
        rest = ", ".join(c.value for c in stack[:-1]) if len(stack) > 1 else ""
        stacks.append(f"({rest}→{top})" if rest else f"({top})")
    return " | ".join(stacks)

def _take_small_sun(state: "GameState", move: Move) -> tuple[list[TileColor], bool]:
    f = _get_factory(state, move.take.factory_id)
    taken, remaining = f.take_from_sun(move.take.color)

    # Spieler legt übrige Steine in gewählter Reihenfolge auf Moon
    moon_log = None
    if remaining:
        f.place_on_moon(move.take.moon_order)
        moon_log = f"🌙 F{f.factory_id} Mond-Stapel: {_format_moon_stacks(f)}"

    # Bonus-Chip aufdecken wenn Fabrik jetzt komplett leer
    chip_log = None
    if f.is_fully_empty and f.bonus_chip and not f.bonus_chip_revealed:
        f.bonus_chip_revealed = True
        chip_log = f"🎴 F{f.factory_id}: Bonusplättchen aufgedeckt!"

    return taken, False, moon_log, chip_log   # kleine Fabrik hat keinen Startspieler-Marker


def _take_small_moon(state: "GameState", move: Move) -> tuple[list[TileColor], bool]:
    f = _get_factory(state, move.take.factory_id)
    taken = f.take_from_moon(move.take.color)

    # Mondstapel nach Entnahme loggen
    remaining_str = _format_moon_stacks(f)
    state.log_event(
        f"🌙 F{f.factory_id} Mond-Stapel nach Entnahme: {remaining_str}"
    )

    # Bonus-Chip aufdecken wenn Fabrik jetzt komplett leer
    if f.is_fully_empty and f.bonus_chip and not f.bonus_chip_revealed:
        f.bonus_chip_revealed = True
        state.log_event(f"🎴 F{f.factory_id}: Bonusplättchen aufgedeckt!")

    return taken, False


def _take_large_sun(state: "GameState", move: Move) -> tuple[list[TileColor], bool]:
    lf = state.large_factory
    taken, remaining, got_marker = lf.take_from_sun(move.take.color)

    # Übrige Steine landen im Moon-Pool der großen Fabrik
    if remaining:
        lf.add_to_moon(remaining)

    return taken, got_marker


def _take_large_moon(state: "GameState", move: Move) -> tuple[list[TileColor], bool]:
    lf = state.large_factory
    taken, got_marker = lf.take_from_moon(move.take.color)
    return taken, got_marker


# ---------------------------------------------------------------------------
# Startspieler-Marker
# ---------------------------------------------------------------------------

def _apply_first_player_marker(state: "GameState") -> None:
    """
    Vergibt den Startspieler-Marker an den aktiven Spieler.
    Laut Regelwerk S.5: dediziertes −2-Feld, NICHT auf der Strafleiste.
    Die −2 Punkte werden am Rundenende via score_penalty() abgezogen.
    """
    player = state.active_player
    player.holds_first_player_marker = True
    state.first_player_next_round = state.current_player
    state.log_event(
        f"🏁 {player.name}: Startspielerstein genommen (−2 Pkt am Rundenende → aktuell {player.score} Pkt)"
    )


# ---------------------------------------------------------------------------
# Place
# ---------------------------------------------------------------------------

def _execute_place(
    state: "GameState",
    tiles: list[TileColor],
    color: TileColor,
    row_index: int,
) -> None:
    """
    Legt die Steine auf die Musterreihe oder Strafleiste.
    Überschuss geht auf die Strafleiste, der Rest vom Überschuss in den Turm.
    """
    player = state.active_player

    if row_index == -1:
        # Direkt auf Strafleiste
        _add_to_penalty(state, tiles)
        return

    row = player.pattern_lines[row_index]
    overflow = row.add_tiles(tiles)

    if overflow:
        _add_to_penalty(state, overflow)


def _add_to_penalty(
    state: "GameState",
    tiles: list[TileColor],
) -> None:
    """
    Legt Steine auf die Strafleiste des aktiven Spielers.
    Steine die nicht mehr drauf passen (max 4) gehen in den Turm.
    """
    player = state.active_player
    before = len(player.broken_tiles)
    to_tower = player.add_broken(tiles)
    after = len(player.broken_tiles)
    if after > before:
        pen_vals = [-1,-2,-3,-4]
        new_slots = [pen_vals[i] for i in range(before, after)]
        state.log_event(
            f"⚠️  {player.name}: {len(tiles)}× auf Strafleiste "
            f"(Slots {before+1}–{after}, {sum(new_slots)} Pkt Strafe)"
        )
    if to_tower:
        state.tower.add(to_tower)
        state.log_event(
            f"⚠️  {player.name}: {len(to_tower)} Stein(e) → Turm (Strafleiste voll)"
        )


# ---------------------------------------------------------------------------
# Hilfsfunktionen
# ---------------------------------------------------------------------------

def _get_factory(state: "GameState", factory_id: int):
    for f in state.factories:
        if f.factory_id == factory_id:
            return f
    raise ValueError(f"Fabrik {factory_id} nicht gefunden.")

File: rust/src/execution.rs
//! Zugausführung (Drafting) — Port von engine/execution.py.
//!
//! Setzt voraus, dass der Move bereits validiert wurde. Loggt jeden Zug
//! ausführlich (wie die Python-Engine) — der Spiel-Log wird zum Debuggen
//! gebraucht (Steinzüge, Mond-Stapel, Chip-Aufdeckungen, Marker, Strafleiste).

use crate::factory::Factory;
use crate::moves::{Move, TakeSource};
use crate::state::GameState;
use crate::tile::TileColor;

pub fn execute_move(state: &mut GameState, m: &Move) {
    if m.is_global_moon_take() {
        execute_moon_take(state, m.take.color, m.place.row_index);
        return;
    }

    let (taken, got_marker, pending) = execute_take(state, m);
    if got_marker {
        apply_first_player_marker(state);
    }

    // Aktions-Log VOR dem Place — danach Füllstand in denselben Eintrag patchen,
    // damit etwaige Strafleisten-Warnungen aus execute_place danach erscheinen.
    let pi = state.current_player;
    let name = state.players[pi].name.clone();
    let dest = dest_label(m.place.row_index);
    let src = src_label(m);
    state.log_event(format!(
        "☀️  {name}: {}× {} von {} → {}",
        taken.len(),
        m.take.color.value(),
        src,
        dest
    ));
    let idx = state.log.len() - 1;

    execute_place(state, &taken, m.place.row_index);

    if m.place.row_index >= 0 {
        let row = &state.players[pi].pattern_lines[m.place.row_index as usize];
        let fill = format!(" [{}/{}]", row.tiles.len(), row.capacity());
        state.log[idx].push_str(&fill);
    }

    for line in pending {
        state.log_event(line);
    }
}

fn find_factory_idx(state: &GameState, factory_id: usize) -> usize {
    state
        .factories
        .iter()
        .position(|f| f.factory_id == factory_id)
        .expect("Fabrik nicht gefunden")
}

/// Deckt das Bonusplättchen auf, falls die Fabrik jetzt komplett leer ist.
/// Gibt true zurück, wenn gerade aufgedeckt wurde.
fn reveal_chip_if_empty(state: &mut GameState, fidx: usize) -> bool {
    let f = &mut state.factories[fidx];
    if f.is_fully_empty() && f.bonus_chip.is_some() && !f.bonus_chip_revealed {
        f.bonus_chip_revealed = true;
        return true;
    }
    false
}

/// Formatiert alle Mondstapel einer Fabrik als lesbaren String ([unten → oben]).
fn format_moon_stacks(f: &Factory) -> String {
    if f.moon_stacks.is_empty() {
        return "leer".into();
    }
    let stacks: Vec<String> = f
        .moon_stacks
        .iter()
        .map(|stack| match stack.last() {
            None => "(?)".into(),
            Some(top) if stack.len() > 1 => {
                let rest: Vec<&str> = stack[..stack.len() - 1].iter().map(|c| c.value()).collect();
                format!("({}→{})", rest.join(", "), top.value())
            }
            Some(top) => format!("({})", top.value()),
        })
        .collect();
    stacks.join(" | ")
}

fn dest_label(row_index: i32) -> String {
    if row_index >= 0 {
        format!("Reihe {}", row_index + 1)
    } else {
        "Strafleiste".into()
    }
}

fn src_label(m: &Move) -> String {
    match m.take.factory_id {
        Some(id) => format!("F{id}"),
        None => "GF".into(),
    }
}

/// Führt den Take-Teil aus. Gibt (genommene Steine, Marker, ausstehende Logs)
/// zurück — die Logs (Mond-Stapel, Chip-Aufdeckung) werden nach dem Aktions-Log
/// geschrieben.
fn execute_take(state: &mut GameState, m: &Move) -> (Vec<TileColor>, bool, Vec<String>) {
    let mut pending: Vec<String> = Vec::new();
    match m.take.source {
        TakeSource::SmallFactorySun => {
            let fid = m.take.factory_id.expect("small sun braucht factory_id");
            let fidx = find_factory_idx(state, fid);
            let (taken, remaining) = state.factories[fidx]
                .take_from_sun(m.take.color)
                .expect("validierter Zug");
            if !remaining.is_empty() {
                state.factories[fidx].place_on_moon(m.take.moon_order.clone());
                pending.push(format!(
                    "🌙 F{fid} Mond-Stapel: {}",
                    format_moon_stacks(&state.factories[fidx])
                ));
            }
            if reveal_chip_if_empty(state, fidx) {
                pending.push(format!("🎴 F{fid}: Bonusplättchen aufgedeckt!"));
            }
            (taken, false, pending)
        }
        TakeSource::SmallFactoryMoon => {
            let fid = m.take.factory_id.expect("small moon braucht factory_id");
            let fidx = find_factory_idx(state, fid);
            let taken = state.factories[fidx]
                .take_from_moon(m.take.color)
                .expect("validierter Zug");
            pending.push(format!(
                "🌙 F{fid} Mond-Stapel nach Entnahme: {}",
                format_moon_stacks(&state.factories[fidx])
            ));
            if reveal_chip_if_empty(state, fidx) {
                pending.push(format!("🎴 F{fid}: Bonusplättchen aufgedeckt!"));
            }
            (taken, false, pending)
        }
        TakeSource::LargeFactorySun => {
            let (taken, remaining, marker) = state
                .large_factory
                .take_from_sun(m.take.color)
                .expect("validierter Zug");
            if !remaining.is_empty() {
                state.large_factory.add_to_moon(&remaining);
            }
            (taken, marker, pending)
        }
        TakeSource::LargeFactoryMoon => {
            let (taken, marker) = state
                .large_factory
                .take_from_moon(m.take.color)
                .expect("validierter Zug");
            (taken, marker, pending)
        }
    }
}

/// Aktion C: nimmt alle obersten Fliesen der Farbe vom Mondbereich ALLER
/// Manufakturen (klein + groß) gleichzeitig.
fn execute_moon_take(state: &mut GameState, color: TileColor, row_index: i32) {
    let mut taken: Vec<TileColor> = Vec::new();
    let mut sources: Vec<(String, usize)> = Vec::new();
    let mut pending: Vec<String> = Vec::new();

    let n_factories = state.factories.len();
    for fidx in 0..n_factories {
        if state.factories[fidx].moon_top_colors().contains(&color) {
            if let Ok(tiles) = state.factories[fidx].take_from_moon(color) {
                let c = tiles.len();
                taken.extend(tiles);
                if c > 0 {
                    sources.push((format!("F{}", state.factories[fidx].factory_id), c));
                }
            }
            pending.push(format!(
                "🌙 F{} Mond-Stapel: {}",
                state.factories[fidx].factory_id,
                format_moon_stacks(&state.factories[fidx])
            ));
            if reveal_chip_if_empty(state, fidx) {
                pending.push(format!(
                    "🎴 F{}: Bonusplättchen aufgedeckt!",
                    state.factories[fidx].factory_id
                ));
            }
        }
    }

    let mut got_marker = false;
    if state.large_factory.moon_colors().contains(&color) {
        if let Ok((tiles, marker)) = state.large_factory.take_from_moon(color) {
            let c = tiles.len();
            taken.extend(tiles);
            got_marker = marker;
            if c > 0 {
                sources.push(("GF".into(), c));
            }
            let pool = &state.large_factory.moon_pool;
            let pool_display = if pool.is_empty() {
                "leer".into()
            } else {
                format!("({})", pool.iter().map(|c| c.value()).collect::<Vec<_>>().join(", "))
            };
            pending.push(format!("🌙 GF Moon-Pool: {pool_display}"));
        }
    }

    assert!(!taken.is_empty(), "Aktion C ohne Mond-Fliesen (validiert?)");

    if got_marker {
        apply_first_player_marker(state);
    }

    let pi = state.current_player;
    let name = state.players[pi].name.clone();
    let dest = dest_label(row_index);
    let src_detail = sources.iter().map(|(_, c)| c.to_string()).collect::<Vec<_>>().join("+");
    let src_labels = sources.iter().map(|(l, _)| l.clone()).collect::<Vec<_>>().join(", ");
    state.log_event(format!(
        "🌙 {name}: {} ({src_detail})× {} von {src_labels} → {dest}",
        taken.len(),
        color.value()
    ));
    let idx = state.log.len() - 1;

    execute_place(state, &taken, row_index);

    if row_index >= 0 {
        let row = &state.players[pi].pattern_lines[row_index as usize];
        state.log[idx].push_str(&format!(" [{}/{}]", row.tiles.len(), row.capacity()));
    }

    for line in pending {
        state.log_event(line);
    }
}

fn apply_first_player_marker(state: &mut GameState) {
    let pi = state.current_player;
    state.players[pi].holds_first_player_marker = true;
    state.first_player_next_round = pi;
    let (name, score) = {
        let p = &state.players[pi];
        (p.name.clone(), p.score)
    };
    state.log_event(format!(
        "🏁 {name}: Startspielerstein genommen (−2 Pkt am Rundenende → aktuell {score} Pkt)"
    ));
}

fn execute_place(state: &mut GameState, tiles: &[TileColor], row_index: i32) {
    let pi = state.current_player;
    if row_index == -1 {
        add_to_penalty(state, tiles);
        return;
    }
    let overflow = state.players[pi].pattern_lines[row_index as usize].add_tiles(tiles);
    if !overflow.is_empty() {
        add_to_penalty(state, &overflow);
    }
}

fn add_to_penalty(state: &mut GameState, tiles: &[TileColor]) {
    let pi = state.current_player;
    let before = state.players[pi].broken_tiles.len();
    let to_tower = state.players[pi].add_broken(tiles);
    let after = state.players[pi].broken_tiles.len();

    if after > before {
        let pen_vals = [-1, -2, -3, -4];
        let sum: i32 = (before..after).map(|i| pen_vals[i]).sum();
        let name = state.players[pi].name.clone();
        state.log_event(format!(
            "⚠️  {name}: {}× auf Strafleiste (Slots {}–{}, {sum} Pkt Strafe)",
            tiles.len(),
            before + 1,
            after
        ));
    }
    if !to_tower.is_empty() {
        let name = state.players[pi].name.clone();
        let n = to_tower.len();
        state.tower.add(&to_tower);
        state.log_event(format!("⚠️  {name}: {n} Stein(e) → Turm (Strafleiste voll)"));
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::moves::{Move, PlaceAction, TakeAction, TakeSource};
    use crate::state::setup_new_game;
    use crate::validation::{generate_valid_moves, validate_move};
    use rand::rngs::StdRng;
    use rand::SeedableRng;

    fn names() -> [String; 2] {
        ["P1".into(), "P2".into()]
    }

    #[test]
    fn generated_moves_are_all_valid_and_nonempty() {
        let mut rng = StdRng::seed_from_u64(99);
        let s = setup_new_game(names(), 0, &mut rng);
        let moves = generate_valid_moves(&s);
        assert!(!moves.is_empty());
        for m in &moves {
            assert!(validate_move(&s, m).is_none(), "ungültiger Zug generiert: {m:?}");
        }
    }

    #[test]
    fn small_sun_take_fills_row_and_moon() {
        let mut rng = StdRng::seed_from_u64(5);
        let mut s = setup_new_game(names(), 0, &mut rng);
        let f = &s.factories[0];
        let color = f.sun_colors()[0];
        let remaining: Vec<TileColor> =
            f.sun_tiles.iter().copied().filter(|&t| t != color).collect();
        let n_color = f.sun_tiles.iter().filter(|&&t| t == color).count();
        let m = Move {
            take: TakeAction {
                source: TakeSource::SmallFactorySun,
                color,
                factory_id: Some(1),
                moon_order: remaining.clone(),
            },
            place: PlaceAction { row_index: 5 },
        };
        assert!(validate_move(&s, &m).is_none());
        let log_before = s.log.len();
        execute_move(&mut s, &m);

        assert!(s.factories[0].sun_is_empty());
        assert_eq!(s.factories[0].moon_is_empty(), remaining.is_empty());
        let row = &s.players[0].pattern_lines[5];
        assert_eq!(row.tiles.len(), n_color);
        assert_eq!(row.color, Some(color));
        // Aktions-Log wurde geschrieben (mind. 1 neuer Eintrag, mit Füllstand).
        assert!(s.log.len() > log_before);
        assert!(s.log.iter().any(|l| l.contains("☀️") && l.contains("F1")));
    }

    #[test]
    fn floor_placement_and_marker_from_large() {
        let mut rng = StdRng::seed_from_u64(11);
        let mut s = setup_new_game(names(), 0, &mut rng);
        let color = s.large_factory.sun_colors()[0];
        let m = Move {
            take: TakeAction {
                source: TakeSource::LargeFactorySun,
                color,
                factory_id: None,
                moon_order: Vec::new(),
            },
            place: PlaceAction { row_index: -1 },
        };
        execute_move(&mut s, &m);
        assert!(s.players[0].holds_first_player_marker);
        assert_eq!(s.first_player_next_round, 0);
        assert!(!s.players[0].broken_tiles.is_empty());
        assert!(s.large_factory.sun_is_empty());
        // Marker- und Strafleisten-Log vorhanden.
        assert!(s.log.iter().any(|l| l.contains("🏁")));
        assert!(s.log.iter().any(|l| l.contains("Strafleiste")));
    }
}
